lib/nmatrix/math.rb

```diff
@@ -623,6 +623,31 @@ def pow n
     product
   end
 
+  def get_rank(workspace_size=1)
+    sigmas = self.gesvd[1].to_a.flatten
+    eps = NMatrix::FLOAT64_EPSILON
+
+    # epsilon depends on the width of the number
+    if (self.dtype == :float32 || self.dtype == :complex64) 
+    eps = NMatrix::FLOAT32_EPSILON
+    end
+
+    tol = self.shape.max * sigmas.max * eps
+
+    sigmas.map { |x| x > tol ? 1 : 0 }.reduce(:+)
+  end
+
+  def full_rank?
+    # if it's a square matrix, save some computation time
+    # and just check if determinant is > 0
+    return self.det != 0 if self.shape[0] == self.shape[1]
```

Comment:
`det` fails to be zero for certain singular square matrices, like these two :-

```
[18] pry(main)> NMatrix.new([3,3], [1.0,2.0,1.0,-2.0,-3.0,1.0,3.0,5.0,0]).det
=> 1.3322676295501884e-15
[19] pry(main)> NMatrix.new([3,3], [1.0,2.0,3,4,5,6,7,8,9]).det
=> 6.661338147750939e-16
```

I have no idea why this happens to be honest but it makes the determinant a bit imprecise to use in the general case. (`det` only approximates using decompositions, and `det_exact`is only implemented for 3x3 or smaller square matrices)